Public/Src/Engine/Processes/ProcessDumper.cs

```diff
@@ -254,53 +268,54 @@ public static bool TryDumpProcessAndChildren(int parentProcessId, string dumpDir
         /// <exception cref="BuildXLException">May throw a BuildXLException on failure</exception>
         internal static List<KeyValuePair<string, int>> GetProcessTreeIds(int parentProcessId, int maxTreeDepth)
         {
-            List<KeyValuePair<string, int>> result = new List<KeyValuePair<string, int>>();
-
-            var maybeProcess = TryGetProcesById(parentProcessId);
-            if (!maybeProcess.Succeeded)
-            {
-                maybeProcess.Failure.Throw();
-            }
-
-            Process p = maybeProcess.Result;
-            result.Add(new KeyValuePair<string, int>("1_" + p.ProcessName + ".exe", p.Id));
-            foreach (var child in GetChildProcessTreeIds(parentProcessId, maxTreeDepth - 1))
-            {
-                result.Add(new KeyValuePair<string, int>("1_" + child.Key, child.Value));
-            }
-
-            return result;
+            return GetChildProcessTreeIds(parentProcessId, maxTreeDepth, isRootQuery: true);
         }
 
-        internal static List<KeyValuePair<string, int>> GetChildProcessTreeIds(int parentProcessId, int maxTreeDepth)
+        private static List<KeyValuePair<string, int>> GetChildProcessTreeIds(int idToQuery, int maxTreeDepth, bool isRootQuery)
         {
             List<KeyValuePair<string, int>> result = new List<KeyValuePair<string, int>>();

```

Comment:
May be not related to this change. Instead of creating a new instance of `List` for each call of `GetChildProcessTreeIds`,  I'm just wondering why `result` is not an argument of `GetChildProcessTreeIds`, and is used to populate the processes?

Reply:
I guess it's due to the logic in the recursion. The outer list is populated with a modified version of the inner recursive call. That could be reworked to have inner call get passed the state from the outer context so the full name could be constructed by the inner recursive call. That would indeed be more efficient.

That said, this code is not performance critical and it works. I don't think it's worth the effort

---
In reply to: [336642280](https://github.com/microsoft/BuildXL/pull/1061#discussion_r336642280) [](ancestors = 336642280)